setup.py

```diff
@@ -1,8 +1,163 @@
+        for d in v.type.decls:
+            if not isinstance(d.type.type, IdentifierType):
+                continue
+            dn = d.name
+            tname = " ".join(d.type.type.names)
+            if tname not in tm:
+                continue
+            fields.append((dn, tm[tname]))
+        return fields
+
+    def get_array_fields(self, v, tm):
+        fields = list()
+        for d in v.type.decls:
+            if not isinstance(d.type, ArrayDecl):
+                continue
+            tname = " ".join(d.type.type.type.names)
+            if tname not in tm:
+                continue
+            fields.append(tm[tname])
+        return fields
+
+    def generate_mappings(self, ast, tm):
+        mapping = dict()
+        element_mapping = dict()
+        for v in ast:
+            if not isinstance(v, Decl) or not isinstance(v.type, Struct):
+                continue
+            fields = self.get_fields(v, tm)
+            for f in fields:
+                mapping[(v.type.name, f[0])] = f[1]
+            afields = self.get_array_fields(v, tm)
+            for af in afields:
+                element_mapping[v.type.name] = af
+        return (mapping, element_mapping)
+
+    def get_mappings(self):
+        pk = pkgconfig.parse("tss2-esys")
+        header_path = None
+        for ip in pk["include_dirs"]:
+            hp = os.path.join(ip, "tss2_tpm2_types.h")
+            if os.path.isfile(hp):
+                header_path = hp
+                break
+            hp = os.path.join(ip, "tss2", "tss2_tpm2_types.h")
+            if os.path.isfile(hp):
+                header_path = hp
+                break
+        if header_path is None:
+            raise RuntimeError(
+                f"unable to find tss2_tpm2_types.h in {pk['include_dirs']}"
+            )
+        pdata = preprocess_file(header_path)
+        parser = c_parser.CParser()
+        ast = parser.parse(pdata, "tss2_tpm2_types.h")
+        tm = self.get_types(ast)
+        return self.generate_mappings(ast, tm)
+
+    def run(self):
+        super().run()
+        type_map, element_type_map = self.get_mappings()
+        mstr = "_type_map = {\n"
+        for k, v in type_map.items():
+            (t, f) = k
+            mstr = mstr + f'    ("{t}", "{f}"): "{v}",\n'
+        mstr = mstr + "}\n"
+        estr = "_element_type_map = {\n"
+        for k, v in element_type_map.items():
+            estr = estr + f'    "{k}": "{v}",\n'
+        estr = estr + "}\n"
+
+        p = os.path.join(self.build_lib, "tpm2_pytss/internal/type_mapping.py")
+        mout = ""
+        with open(p, "r+t") as tf:
+            for l in tf.readlines():
+                if l == "_type_map = {}\n":
+                    mout = mout + mstr
+                elif l == "_element_type_map = {}\n":
+                    mout = mout + estr
+                else:
+                    mout = mout + l
+            tf.seek(0)
+            tf.truncate(0)
+            tf.write(mout)
```

Comment:
Why do we need to read in the file and then update it? Don't we just take the generated map and dump it to the file no matter what? This would be a file we don't include in the source tree and add to .gitignore.

Reply:
The primary reason I choose that is so we don't need to catch any import exceptions in types.py, but I'm fine with either way

Reply:
I changed it so it will overwrites the whole file